keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,176 @@
+Staging branch is automatically applied to the staging Kubernetes cluster by a push triggered Jenkins job.
+All members of her team file pull requests against the repo to bring changes into staging and production.
+
+Aside from that, Alice and her team got services that directly manage their objects, like controllers and webhooks,
+as well as services that updated labels for some objects.
+
+They want to make use of the new Server Side Apply, so concurrent changes to objects from different actors are possible without unwanted overlap/conflicts.
+
+To do so, assuming the changes in this KEP, Alice would add the newly introduced field to all manifests in their config management repo. The field would be set to e.g. `configManagement` (or their repo name).
+Additionally, she would update the manifests sent by different actors (like controllers) to contain the field with a value like `controllerName`. Note that this might already be the case for some through the defaulting to the user-agent inside the apply and update pathway.
+
+If now the team updates their repo with a change to their manifests, Jenkins again applies them as usual. On the apiserver side though, the fieldManager set in the applied manifest gets used and all fields set now get owned by the `configManagement` fieldManager.
+
+Again as usual other actors step in to change the objects. They as well use their fieldManager value either set in the manifest file they apply, or based on their user-agent/configuration.
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+#### Kustomize
+
+In reference to the above workflow, it would be possible to override `options.fieldManager` for differen Kustomize overlays and therefor separate ownership depending on the actor currently active.
+
+This means, updating single fields by using a Kustomize overlay (like updating images, labels or annotations) as part of a CI/CD pipeline could easily contain the fieldManager information without any changes to Kustomize (or similar tools) itself.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it causes different behavior on the apiserver, but is not reflected when reading the object back from the apiserver.
+
+This can seem unintuitive but should be solved by proper documentation.
+
+## Design Details
+
+### Proposed API Change
+
+```go
+type ObjectMeta struct {
+...
+  // Options used by the apiserver when handling the object.
+  // This field is write-only, non-persisted and optional.
+  Options *MetaOptions `json:"options,omitempty" protobuf:"bytes,18,opt,name=options"`
```

Comment:
For discussion: Suppose we call this "WriteOnlyOptions"? Will we ever have an option that is not write-only?

Reply:
I don't think so.
Another option I could think of right now would be `force` but its write-only as well.